From ad541f63ceafbe6aba31d8dea0d96da47fa055a8 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Patrick=20L=C3=BCdeke?= <patrick@vivida.net>
Date: Wed, 21 Feb 2024 22:17:29 +0100
Subject: [PATCH] Reduce RabbitMQ health check CPU load

On machines with many cores (256 in our case), we were seeing a high CPU
load caused by the RabbitMQ health check and rabbitmq-diagnostics tool.
A previous commit (see related commits) fixed this issue for RabbitMQ
server. This commit fixes the issue for the health check and
rabbitmq-diagnostics tool. With the following Erlang VM adjustments we
saw a drop in CPU Load from >15 to <1.

  RABBITMQ_CTL_ERL_ARGS="+S 2:2"

This pins the rabbitmq-diagnostics tool to two scheduler threads instead
of N, where N is the number of CPU cores.

Closes-Bug: #2054607
Related-Commit: 70f6f8e4c02ed6a8687d2bd714d3fe0b9d04d84a
Related-Bug: #1846467
Change-Id: Ib1bd55862a99dc05c6357fda9d4cd7960fef88c9
---
 ansible/group_vars/all.yml                             |  1 +
 ansible/roles/common/templates/rabbitmq-env.conf.j2    |  2 +-
 ansible/roles/rabbitmq/defaults/main.yml               |  2 +-
 ansible/roles/rabbitmq/templates/rabbitmq-env.conf.j2  |  2 +-
 etc/kolla/globals.yml                                  |  3 ++-
 ...abbitmq-health-check-cpu-load-2690a6e645512e76.yaml | 10 ++++++++++
 6 files changed, 16 insertions(+), 4 deletions(-)
 create mode 100644 releasenotes/notes/rabbitmq-health-check-cpu-load-2690a6e645512e76.yaml

diff --git a/ansible/group_vars/all.yml b/ansible/group_vars/all.yml
index 288b3c3415..ad613dcd1a 100644
--- a/ansible/group_vars/all.yml
+++ b/ansible/group_vars/all.yml
@@ -616,6 +616,7 @@ rabbitmq_management_port: "15672"
 rabbitmq_cluster_port: "25672"
 rabbitmq_epmd_port: "4369"
 rabbitmq_prometheus_port: "15692"
+rabbitmq_ctl_erl_args: "+S 2:2"
 
 redis_port: "6379"
 redis_sentinel_port: "26379"
diff --git a/ansible/roles/common/templates/rabbitmq-env.conf.j2 b/ansible/roles/common/templates/rabbitmq-env.conf.j2
index b78696a209..b1dfe0d40f 100644
--- a/ansible/roles/common/templates/rabbitmq-env.conf.j2
+++ b/ansible/roles/common/templates/rabbitmq-env.conf.j2
@@ -1,2 +1,2 @@
-RABBITMQ_CTL_ERL_ARGS="{% if api_address_family == 'ipv6' %}-proto_dist inet6_tcp {% endif %}"
+RABBITMQ_CTL_ERL_ARGS="{% if api_address_family == 'ipv6' %}-proto_dist inet6_tcp {% endif %}{{ rabbitmq_ctl_erl_args }}"
 export ERL_INETRC=/etc/rabbitmq/erl_inetrc
diff --git a/ansible/roles/rabbitmq/defaults/main.yml b/ansible/roles/rabbitmq/defaults/main.yml
index bf2573fef7..4c336a7e66 100644
--- a/ansible/roles/rabbitmq/defaults/main.yml
+++ b/ansible/roles/rabbitmq/defaults/main.yml
@@ -80,7 +80,7 @@ rabbitmq_extra_volumes: "{{ default_extra_volumes }}"
 # Message-Broker
 ####################
 rabbitmq_pid_file: "/var/lib/rabbitmq/mnesia/rabbitmq.pid"
-rabbitmq_server_additional_erl_args: "+S 2:2 +sbwt none +sbwtdcpu none +sbwtdio none"
+rabbitmq_server_additional_erl_args: "{{ rabbitmq_ctl_erl_args }} +sbwt none +sbwtdcpu none +sbwtdio none"
 # Dict of TLS options for RabbitMQ. Keys will be prefixed with 'ssl_options.'.
 rabbitmq_tls_options: {}
 # To avoid split-brain
diff --git a/ansible/roles/rabbitmq/templates/rabbitmq-env.conf.j2 b/ansible/roles/rabbitmq/templates/rabbitmq-env.conf.j2
index 90d0561c0f..e0faadf88e 100644
--- a/ansible/roles/rabbitmq/templates/rabbitmq-env.conf.j2
+++ b/ansible/roles/rabbitmq/templates/rabbitmq-env.conf.j2
@@ -3,7 +3,7 @@ RABBITMQ_LOG_BASE=/var/log/kolla/{{ project_name }}
 RABBITMQ_DIST_PORT={{ role_rabbitmq_cluster_port }}
 RABBITMQ_PID_FILE={{ rabbitmq_pid_file }}
 RABBITMQ_SERVER_ADDITIONAL_ERL_ARGS="-kernel inetrc '/etc/rabbitmq/erl_inetrc' {% if api_address_family == 'ipv6' %}-proto_dist inet6_tcp {% endif %}{{ rabbitmq_server_additional_erl_args }}"
-RABBITMQ_CTL_ERL_ARGS="{% if api_address_family == 'ipv6' %}-proto_dist inet6_tcp {% endif %}"
+RABBITMQ_CTL_ERL_ARGS="{% if api_address_family == 'ipv6' %}-proto_dist inet6_tcp {% endif %}{{ rabbitmq_ctl_erl_args }}"
 
 export ERL_EPMD_ADDRESS={{ api_interface_address }}
 export ERL_EPMD_PORT={{ role_rabbitmq_epmd_port }}
diff --git a/etc/kolla/globals.yml b/etc/kolla/globals.yml
index 171211b3fa..4795740684 100644
--- a/etc/kolla/globals.yml
+++ b/etc/kolla/globals.yml
@@ -464,7 +464,8 @@ workaround_ansible_issue_8743: yes
 # https://www.rabbitmq.com/runtime.html#busy-waiting
 # The default tells RabbitMQ to always use two cores (+S 2:2),
 # and not to busy wait (+sbwt none +sbwtdcpu none +sbwtdio none):
-#rabbitmq_server_additional_erl_args: "+S 2:2 +sbwt none +sbwtdcpu none +sbwtdio none"
+#rabbitmq_ctl_erl_args: "+S 2:2"
+#rabbitmq_server_additional_erl_args: "{{ rabbitmq_ctl_erl_args }} +sbwt none +sbwtdcpu none +sbwtdio none"
 # Whether to enable TLS encryption for RabbitMQ client-server communication.
 #rabbitmq_enable_tls: "no"
 # CA certificate bundle in RabbitMQ container.
diff --git a/releasenotes/notes/rabbitmq-health-check-cpu-load-2690a6e645512e76.yaml b/releasenotes/notes/rabbitmq-health-check-cpu-load-2690a6e645512e76.yaml
new file mode 100644
index 0000000000..2585e678b0
--- /dev/null
+++ b/releasenotes/notes/rabbitmq-health-check-cpu-load-2690a6e645512e76.yaml
@@ -0,0 +1,10 @@
+---
+fixes:
+  - |
+    Fixes an issue where RabbitMQ health check and rabbitmq-diagnostics
+    cause a high CPU load on systems with many CPU cores.
+features:
+  - |
+    Adds ``rabbitmq_ctl_erl_args`` variable to set additional
+    ``RABBITMQ_CTL_ERL_ARGS`` args. Please see:
+    https://www.rabbitmq.com/runtime.html#configure
